src/Catrobat/Security/FOSUBUserProvider.php

```diff
@@ -0,0 +1,121 @@
+    }
+    //retrieve access token and the ID
+    $property = $this->getProperty($response);
+    $username = $response->getUsername();
+    $service = $response->getResourceOwner()->getName();
+
+    $setter_name = 'set'.ucfirst($service);
+    $setter_id = $setter_name.'Id';
+    $setter_access_token = $setter_name.'AccessToken';
+    $access_token = $response->getAccessToken();
+
+    //Disconnect previous user
+    if (null !== $previousUser = $this->userManager->findUserBy([$property => $username]))
+    {
+      $previousUser->{$setter_id}(null);
+      $previousUser->{$setter_access_token}(null);
+      $this->userManager->updateUser($user);
+    }
+    //username is a unique integer
+    $user->{$setter_id}($username);
+    $user->{$setter_access_token}($access_token);
+    $this->userManager->updateUser($user);
+  }
+
+  public function loadUserByOAuthUserResponse(UserResponseInterface $response): UserInterface
+  {
+    $username = $response->getUsername();
+
+    $user = $this->userManager->findUserBy([$this->getProperty($response) => $username]);
+    $service = $response->getResourceOwner()->getName();
+    //make setter and getter names dynamic so we can use it for more services (Facebook, Google, Github)
+    $setter_name = 'set'.ucfirst($service);
+    $setter_id = $setter_name.'Id';
+    $setter_access_token = $setter_name.'AccessToken';
+    $access_token = $response->getAccessToken();
+    //register new user
+    if (null === $user)
+    {
+      $user = $this->userManager->findUserByEmail($response->getEmail());
+      //if user with the given email doesnt exists create a new user
+      if (null === $user)
+      {
+        /** @var User $user */
+        $user = $this->userManager->createUser();
+        //generate random username for example user12345678, needs to be discussed
+        $user->setUsername($this->createRandomUsername());
+        $user->setEmail($response->getEmail());
+        $user->setEnabled(true);
+        $user->setPassword($this->generateRandomPassword());
+        $user->setOauthUser(true);
+      }
+      $user->{$setter_id}($username);
+      $user->{$setter_access_token}($access_token);
+      $this->userManager->updateUser($user);
+
+      return $user;
+    }
+
+    //if user exists just proceed with hwioauthentication
+    $user = parent::loadUserByOAuthUserResponse($response);
+
+    //update access token
+    $user->{$setter_access_token}($access_token);
+
+    return $user;
+  }
+
+  private function generateRandomPassword(int $length = 32): string
+  {
+    $chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'.
+      '0123456789-=~!@#$%&*()_+,.<>?;:[]{}|';
+
+    $pass = '';
+    $max = strlen($chars) - 1;
+
+    for ($i = 0; $i < $length; ++$i)
+    {
+      try
+      {
+        $pass .= $chars[random_int(0, $max)];
+      }
+      catch (Exception $e)
+      {
+      }
```

Comment:
this catch might be dangerous.
Maybe it is better if we let the application crash in this scenario - or use a different approach for the password in this case?

In the worst case we might get users without passwords. The exception should not happen anyway. So maybe just remove the try/catch